Afft type

burdges opened the pull request:
Superceeds https://github.com/drahnr/rs-ec-perf/pull/10

burdges committed:
Fix sampling

burdges committed:
make mul_table more readable

burdges committed:
reformat walsh code

burdges committed:
unexpand -t 4

burdges committed:
Oops typeos

burdges committed:
b_not_one

burdges committed:
s/MODULO/ONEMASK/

burdges committed:
ReedSolomon::new is a very bad name for a 1/3 thing

burdges committed:
Start abstracting field

burdges committed:
Revert "Fix sampling"

This reverts commit 7aa4dbd00a498ec8009f20adc5c519461d71389d.

burdges committed:
Create and use pre-generated tables

We'll eventually need a seperate tables.rs build target because cargo
dislikes build artifacts appearing outside env!("OUT_DIR") and we
require tables to build other tables.

ref.  https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script

burdges committed:
Correct log and exp table for this f2e16 with the original base, etc.

burdges committed:
Move LOG_WALSH to the field

We've no idea if the LOG_WALSH table can be shrunk for the domain,
but it appears mostly unimportant for us.

burdges committed:
s/_fft_/_afft_/g

Additive FFT

burdges committed:
one commented out option

burdges committed:
Attempt to being making more type safe

burdges committed:
Compute skew factor using field methods

burdges committed:
What is B doing?  And why does it do nothing?

burdges committed:
Improve check that B does ntohing

burdges committed:
Make B: [Multiplier; _]

burdges committed:
Hide B completely inside test

burdges committed:
Use Multiplier in Walsh transform types

burdges committed:
SKEW_FACTOR: [Multiplier; ONEMASK as usize]

drahnr committed:
Replace codewords with additive representation

drahnr committed:
fix make change endianness independent

drahnr committed:
remove some circular generation dependency insanity

Tables are generated whenever the field impl changes, not on manual
copy, they also live under OUT_DIR right now.

drahnr committed:
chore: cargo fmt + fix

burdges committed:
tmp

burdges committed:
Improve field table handling

burdges committed:
more field cleanup

burdges committed:
tmp

burdges committed:
use static_init

saves 2ns per access using this crate and unsafe

burdges committed:
s/mem_cpy/copy_from_slice/

burdges committed:
use non-polkadot terminology

s/derive_from_validator_count/derive_from_third_plus_epsilon/
s/validator_count/real_n/
s/is_power_of_2/is_nonzero_power_of_2/

burdges committed:
misc

burdges committed:
s/ReedSolomon/CheckedParams/

drahnr commented on src/novel_poly_basis.rs:
```diff
@@ -12,100 +12,29 @@ use super::*;
-
+include!("params.rs");
 include!("encode.rs");
 include!("decode.rs");
```
This should be avoided at all cost except for the table generation.

burdges replied:
They'll likely become separate modules eventually.  I'm just not ready to settle on exactly what the field functionality will be, but I'm kinda done working on them being in the same file.

If you've come concrete objection like it messes up RLS then I'd consider delaying merging this part of 11 into 12 but right now them all living in one big sloppy file is only making it harder for the final structure to emerge.

burdges committed:
Add Ming-Shing Chen's C code for GF(16) and GF(256)

From https://github.com/fast-crypto-lab/rainbow-submission-round2/tree/master/Alternative_Implementation/ssse3/IIIc_Classic

burdges committed:
todo

burdges committed:
copy c tables into rust

burdges committed:
Cantor basis test taht does not work right now

burdges committed:
Add back f256

burdges committed:
make f256 build

drahnr commented on src/f256.rs:
```diff
@@ -0,0 +1,35 @@
+include!("f_log_mul.rs");
+
+#[cfg(table_bootstrap_complete)]
+include!("afft.rs");
```
No!

burdges replied:
I do think `pub(crate) struct AdditiveFFT<F: BinaryField>` could exist without this since `AdditiveFFT::afft`/`inverse_afft` require relatively little too, at least initially though maybe later maybe they require macros.

`AdditiveFFT::initialize` should definitely be a macro or use `include!` though because we'll yank the field representation out from under it I think.  If you read my comments in https://github.com/drahnr/rs-ec-perf/pull/11/files#diff-f3f3b9070bc5ad37ca82632b2043fe553dcade424b615a4138478e105dcdde61R253-R275 then you'll see we do not currently understand enough to change or abstract it though.

drahnr commented on src/f256.rs:
```diff
@@ -0,0 +1,35 @@
+pub const FIELD_BITS: usize = 8;
+pub const FIELD_SIZE: usize = 1_usize << FIELD_BITS;
+
+include!("f_log_mul.rs");
```
No!

burdges replied:
It seemed cleaner than macros.  lol

burdges committed:
b_is_one test

burdges committed:
Extremely preliminate degree two extension multiplier form

burdges committed:
make cantor_basis test work

burdges committed:
abstract cantor_basis test

drahnr commented:
@burdges  _Most_ of the changes are now part of master, minus the `C` files. Feel free to re-add them. This was a best effort merge and tried to not introduce any abstractions at all.

Closing this.
